### PR TITLE
Replacing of "get distance" to "get distance squared"

### PR DESCRIPTION
**Pull request by MekBoss**
Slightly misleading title.
What it actually do, is replaced most checks, which involves distance measurement to check is actor closer/further than some specific distance, to use distance squared instead which increase performance. Especially for "NoiseMarkSector" function with non zero "hear distance" argument.

### CODE REVIEW

**Comment by alexey-lysiuk** [comment]
> which increase performance

Do you have numbers to compare? If so, please provide them along with test procedures.

**Comment by MajorCooke** [comment]
I'm certainly thinking about giving this a try in QZDoom.

**Comment by nashmuhandes** [comment]
@MekBoss should you also apply the same treatment for CheckIfSeen, CheckSightOrRange and CheckRange?

**Comment by MajorCooke** [comment]
I personally don't think this needs 'numbers' per say, but if you want a good place to test it... Well, try nuts.wad or okuplok.wad since this hits all the right spots involving monster chasing and movement.

**Comment by madame-rachelle** [comment]
> 
> 
> > which increase performance
> 
> Do you have numbers to compare? If so, please provide them along with test procedures.

I'm actually more worried about breaking stuff than the actual numbers in this. Unless there's something in this code base that I am unaware of, in general the less square-root functions you do, it's always better, almost universally. And whenever I have coded a distance function I have always checked the squared distance, myself, not even bothering with square roots for conditional checks.

I'm surprised that MekBoss was able to find so many opportunities to remove them in the first place.

**Comment by MajorCooke** [comment]
There is already a portal-aware variant made, that was just done for consistency by him to replace the functions that were using the non-portal aware version wherever they were called.

**Comment by drfrag666** [comment]
Is there a way to detect if the map has portals and use the non portal aware version for increased performance? And would it make a difference?

File: src/playsim/p_actionfunctions.cpp [hunk 1]
```diff
@@ -2458,13 +2459,14 @@ DEFINE_ACTION_FUNCTION(AActor, CheckIfTargetInLOS)
 	{
 		ACTION_RETURN_BOOL(false);
 	}
-	double distance = self->Distance3D(target);
+	double distance_squared = self->Distance3DSquared(target);
 
-	if (dist_max && (distance > dist_max))
+	if (dist_max && (distance_squared > (dist_max * dist_max) ) )
```
Comment (edward-san) [inline src/playsim/p_actionfunctions.cpp:2464]: The behavior is changed if `dist_max` is negative. Please adjust this.

File: src/playsim/p_actionfunctions.cpp [hunk 2]
```diff
@@ -2458,13 +2459,14 @@ DEFINE_ACTION_FUNCTION(AActor, CheckIfTargetInLOS)
...
 	{
 		ACTION_RETURN_BOOL(false);
 	}
-	if (dist_close && (distance < dist_close))
+
+	if (dist_close && (distance_squared < (dist_close * dist_close) ) )
```
Comment (edward-san) [inline src/playsim/p_actionfunctions.cpp:2469]: Same here as the comment for `dist_max`.

**Comment by MekBoss** [comment]
@alexey-lysiuk 
> Do you have numbers to compare? If so, please provide them along with test procedures.
Numbers become significant only in edge cases, if someone run Russian Overkill with/or some slaughter map/complex map architecture like several new Sunder maps, for example. With "default" maps its microseconds only optimization, even if I spam with a_alertmonster 1000 times in single tick.

@coelckers 
> The new function is not portal-aware, it should mimic Distance2D, minus the square root.
But original function wasnt portal aware either. There was this
https://github.com/coelckers/gzdoom/blob/ab8d2ef62d33eeebc2799e8d8cd39fe25bbcc199/src/playsim/actor.h#L866
It was used here and in the next line
https://github.com/coelckers/gzdoom/blob/ab8d2ef62d33eeebc2799e8d8cd39fe25bbcc199/src/playsim/p_map.cpp#L1306
While I can rewrite it, wasnt it broken all this time? So I would know should I touch it or revert it back to previous state, to not break compatibility with something.

**Comment by alexey-lysiuk** [comment]
> Numbers become significant only in edge cases, if someone run Russian Overkill with/or some slaughter map/complex map architecture like several new Sunder maps, for example. With "default" maps its microseconds only optimization, even if I spam with a_alertmonster 1000 times in single tick.

What's the point of this then? Without a performance comparison, any optimization is useless.
Before making a pull request, you need to benchmark it, and post the results.
I cannot say for others, but I won't make any decisions without seeing such comparison.
If the change saves 1 microsecond per tick in the worst case scenario, I would leave everything as is.

**Comment by MekBoss** [comment]
> What's the point of this then? Without a performance comparison, any optimization is useless.
> Before making a pull request, you need to benchmark it, and post the results.
> I cannot say for others, but I won't make any decisions without seeing such comparison.
> If the change saves 1 microsecond per tick in the worst case scenario, I would leave everything as is.

My bad, I mean **MILI**second (1/1000)

**Comment by coelckers** [comment]
> What's the point of this then? Without a performance comparison, any optimization is useless.
> Before making a pull request, you need to benchmark it, and post the results.
> I cannot say for others, but I won't make any decisions without seeing such comparison.
> If the change saves 1 microsecond per tick in the worst case scenario, I would leave everything as is.

Agreed. In this case, if performance is the issue, it makes a lot more sense to convert the functions to native with direct entry points. The savings of this would be just as significant as the avoidance of square roots, but far less invasive.
But without some actual numbers of realistic use cases it looks like a change for the sake of changing it.

**Comment by MajorCooke** [comment]
I've merged this into QZDoom for testing.

**Comment by MekBoss** [comment]
Test map
[sound_test.zip](https://github.com/coelckers/gzdoom/files/4618219/sound_test.zip)
To use load it with Release Gzdoom version, type into console "map sound_test"  and then "give alerter".
There are also two cvars "times_alert" and "distance_alert", which controls amount of times a_alertmonsters would called for one "alerter" item and distance from player inside of which monsters would be "alerted".

By default they have value 10000 each. On test map inside of this range located 3000 monsters, so this is 3000 * 10000 = 30.000.000 square root/multiplication operations.

Numbers.
For my machine processing of 10000 calls of a_alertmonsters with "squared length" takes around 400 +-20 ms, rounded up.
Processing of same 10000 calls with "length root" takes around 2200 +- 100 ms, rounded up too.
More than 5 times faster, for me at least. Yaaaaaaaay :)

**Comment by MajorCooke** [comment]
@MekBoss I do hope you change what's requested of you especially with the code, because it seems like FloatSpeed is no longer working.

**Comment by MekBoss** [comment]
Edwards requested changes have nothing to do with float speed property, since it in CheckIfTargetInLOS function. And only things that use float speed property, and that I touched, located in p_enemy.cpp and c_map.cpp.
And I cant trigger buggy behavior both in PR version and release 4.3.3. Example please?

**Comment by MajorCooke** [comment]
Go to a map with a large height between it and the ground. Summon a cacodemon near the roof, then drop to the ground. Watch as the cacodemon does not float down after you.

**Comment by MekBoss** [comment]
Okay, fixed it.

Still wait for Graf response about the function.
So again, original function, "squared" copy of which I created, wasnt portal-aware in the first place. Should I anyway make it portal-aware?

File: src/playsim/p_actionfunctions.cpp [hunk 3]
```diff
@@ -3349,7 +3351,7 @@ DEFINE_ACTION_FUNCTION(AActor, A_WolfAttack)
 
 	// Target can dodge if it can see enemy
 	DAngle angle = absangle(self->target->Angles.Yaw, self->target->AngleTo(self));
-	bool dodge = (P_CheckSight(self->target, self) && angle < 30. * 256. / 360.);	// 30 byteangles ~ 21�
+	bool dodge = (P_CheckSight(self->target, self) && angle < 30. * 256. / 360.);	// 30 byteangles ~ 21�
```
Comment (edward-san) [inline src/playsim/p_actionfunctions.cpp:3354]: Just a nit: which editor was used? IMHO this should be dropped.

File: src/playsim/p_enemy.cpp [hunk 4]
```diff
@@ -140,8 +140,9 @@ static void NoiseMarkSector(sector_t *sec, AActor *soundtarget, bool splash, AAc
 	// [RH] Set this in the actors in the sector instead of the sector itself.
 	for (AActor *actor = sec->thinglist; actor != NULL; actor = actor->snext)
 	{
-		if (actor != soundtarget && (!splash || !(actor->flags4 & MF4_NOSPLASHALERT)) &&
-			(!maxdist || (actor->Distance2D(emitter) <= maxdist)))
+		if(actor != soundtarget && (!splash || !(actor->flags4 & MF4_NOSPLASHALERT)) &&
+			(!maxdist ||
+			(maxdist >= 0 && actor->Distance2DSquared(emitter) <= (maxdist * maxdist) ) ) )
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:145]:
```suggestion
			(!maxdist || (actor->Distance2DSquared(emitter) <= maxdist * abs(maxdist))))
```

File: src/playsim/p_enemy.cpp [hunk 5]
```diff
@@ -1163,12 +1169,12 @@ int P_IsVisible(AActor *lookee, AActor *other, INTBOOL allaround, FLookExParams
 		fov = allaround ? 0. : 180.;
 	}
 
-	double dist = lookee->Distance2D (other);
-
-	if (maxdist && dist > maxdist)
+	double distance_squared = lookee->Distance2DSquared(other);
+	//square distance here...
+	if (maxdist && maxdist >= 0 && distance_squared > (maxdist * maxdist) )
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:1174]:
```suggestion
	if (maxdist && distance_squared > maxdist * abs(maxdist))
```

File: src/playsim/p_enemy.cpp [hunk 6]
```diff
@@ -1163,12 +1169,12 @@ int P_IsVisible(AActor *lookee, AActor *other, INTBOOL allaround, FLookExParams
...
 		return false;			// [KS] too far
-
-	if (mindist && dist < mindist)
+	//...and here because other number are squared
+	if (mindist && mindist >= 0 && distance_squared < (mindist * mindist) )
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:1177]:
```suggestion
	if (mindist && distance_squared < mindist * abs(mindist))
```

File: src/playsim/p_enemy.cpp [hunk 7]
```diff
@@ -1923,10 +1929,10 @@ DEFINE_ACTION_FUNCTION(AActor, A_LookEx)
 				}
 				else
 				{
-					dist = self->Distance2D (targ);
+					distance_squared = self->Distance2DSquared(targ);
 
 					// [KS] If the target is too far away, don't respond to the sound.
-					if (maxheardist && dist > maxheardist)
+					if (maxheardist && maxheardist > 0 && distance_squared > (maxheardist * maxheardist) )
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:1935]:
```suggestion
					if (maxheardist && distance_squared > maxheardist * abs(maxheardist))
```

File: src/playsim/p_enemy.cpp [hunk 8]
```diff
@@ -1994,10 +2000,10 @@ DEFINE_ACTION_FUNCTION(AActor, A_LookEx)
 			{
 				if (self->flags & MF_AMBUSH)
 				{
-					dist = self->Distance2D (self->target);
+					distance_squared = self->Distance2D (self->target);
 					if (P_CheckSight (self, self->target, SF_SEEPASTBLOCKEVERYTHING) &&
-						(!minseedist || dist > minseedist) &&
-						(!maxseedist || dist < maxseedist))
+						(!minseedist || (minseedist >= 0 && distance_squared > (minseedist * minseedist) ) ) &&
+						(!maxseedist || (maxseedist >= 0 && distance_squared < (maxseedist * maxseedist) ) ) )
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:2006]:
```suggestion
						(!minseedist || distance_squared > minseedist * abs(minseedist)) &&
						(!maxseedist || distance_squared < maxseedist * abs(maxseedist)))
```

File: src/playsim/p_mobj.cpp [hunk 9]
```diff
@@ -2479,11 +2479,17 @@ void P_ZMovement (AActor *mo, double oldfloorz)
...
+			distance_squared = mo->Distance2DSquared(mo->target);
+			//still need delta intact, because it can be negative
+			//and square of a real number always positive
+			delta = mo->target->Center() - mo->Z();
+			//why it multiplies by 3 in original function?
+			double delta_squared = delta * abs(delta) * 3 * 3;
+			if (delta_squared < 0 && distance_squared < -(delta_squared) )
 				mo->AddZ(-mo->FloatSpeed);
-			else if (delta > 0 && dist < (delta*3))
+
```
Comment (edward-san) [inline src/playsim/p_mobj.cpp:2491]: IMHO this emtpy line here can be dropped.

**Comment by MekBoss** [comment]
No?! **abs** work fine for all numeric data types in C++, only in C it require **fabs** for float and simple **abs** for integers.

**Comment by Doom2fan** [comment]
I'm seeing a few suggestions for changes on the PR here that don't make sense - For distance arguments that accept negative numbers, those should be changed to check if the number is greater than 0, not use abs or anything like that. Something like that would make it pass with negative values, while such checks would never pass with the original code.
(
Remember, the calculated distances can never be < 0, so passing in negative values to any max, min, etc. distance arguments that don't have special behaviour will make the checks never pass)

**Comment by edward-san** [comment]
Of course the code can be changed to check if a parameter is nonnegative or not, but IMHO it should be something to do outside of this PR, since 1) it is a potentially mod-breaking change and 2)it is not related to the original intent of "replacing the distance with its squared distance".

**Comment by Doom2fan** [comment]
> 
> 
> Of course the code can be changed to check if a parameter is nonnegative or not, but IMHO it should be something to do outside of this PR, since 1) it is a potentially mod-breaking change and 2) it is not related to the original intent of "replacing the distance with its squared distance".

That's not what I'm saying. Doing it like **this** with abs would be a mod-breaking change. Again, negative values will never pass the checks, because distance is never negative, which means the correct behaviour is to fail such checks if the specified value for the argument is negative.

**Comment by edward-san** [comment]
> That's not what I'm saying. Doing it like **this** with abs would be a mod-breaking change. Again, negative values will never pass the checks, because distance is never negative, which means the correct behaviour is to fail such checks if the specified value for the argument is negative.

Please elaborate it rigorously. Distance is never negative, sure, but the parameters can be and they are legit in modding (for instance, a sensor which changes its scanning range periodically, including a 'blindfold' time, or also a 'indefinitely maximum range' time).

File: src/playsim/p_actionfunctions.cpp [hunk 10]
```diff
@@ -2458,13 +2459,14 @@ DEFINE_ACTION_FUNCTION(AActor, CheckIfTargetInLOS)
 	{
 		ACTION_RETURN_BOOL(false);
 	}
-	double distance = self->Distance3D(target);
+	double distance_squared = self->Distance3DSquared(target);
 
-	if (dist_max && (distance > dist_max))
+	if (dist_max && (dist_max >= 0 && distance_squared > (dist_max * dist_max) ) )
```
Comment (edward-san) [inline src/playsim/p_actionfunctions.cpp:2464]: This comparison is wrong.

```suggestion
	if (dist_max && (dist_max < 0 || distance_squared > dist_max*dist_max))
```

File: src/playsim/p_actionfunctions.cpp [hunk 11]
```diff
@@ -2458,13 +2459,14 @@ DEFINE_ACTION_FUNCTION(AActor, CheckIfTargetInLOS)
...
 	{
 		ACTION_RETURN_BOOL(false);
 	}
-	if (dist_close && (distance < dist_close))
+
```
Comment (edward-san) [inline src/playsim/p_actionfunctions.cpp:2468]: this line can be dropped

File: src/playsim/p_actionfunctions.cpp [hunk 12]
```diff
@@ -2549,16 +2551,16 @@ DEFINE_ACTION_FUNCTION(AActor, CheckIfInTargetLOS)
 		ACTION_RETURN_BOOL(false);
 	}
 
-	double distance = self->Distance3D(target);
+	double distance_squared = self->Distance3DSquared(target);
 
-	if (dist_max && (distance > dist_max))
+	if (dist_max && dist_max >= 0 && (distance_squared > (dist_max * dist_max) ) )
```
Comment (edward-san) [inline src/playsim/p_actionfunctions.cpp:2556]:
```suggestion
	if (dist_max && (dist_max < 0 || distance_squared > dist_max*dist_max))
```

File: src/playsim/p_enemy.cpp [hunk 13]
```diff
@@ -944,16 +945,21 @@ void P_NewChaseDir(AActor * actor)
...
 			bool ismeleeattacker = false;
-			double dist = actor->Distance2D(target);
+
+			double distance_squared = actor->Distance2DSquared(target);
 			if (target->player == NULL)
 			{
-				ismeleeattacker = (target->MissileState == NULL && dist < (target->meleerange + target->radius)*2);
+				//in its own variable because otherwise it looks bad
+				double meele_squared = (target->meleerange + target->radius) * 2 * (target->meleerange + target->radius) * 2;
+				ismeleeattacker = (target->MissileState == NULL && distance_squared < meele_squared);
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:954]: it's better safe than sorry if you do this:

```suggestion
				double meele = (target->meleerange + target->radius)*2;
				ismeleeattacker = (target->MissileState == NULL && (melee > 0 && distance_squared < meele*melee));
```

File: src/playsim/p_enemy.cpp (hunk 5 above)
Comment (edward-san) [inline src/playsim/p_enemy.cpp:1174]:
```suggestion
	if (maxdist && (maxdist < 0 || distance_squared > maxdist*maxdist))
```

File: src/playsim/p_enemy.cpp [hunk 14]
```diff
@@ -1179,7 +1185,7 @@ int P_IsVisible(AActor *lookee, AActor *other, INTBOOL allaround, FLookExParams
 		{
 			// if real close, react anyway
 			// [KS] but respect minimum distance rules
-			if (mindist || dist > lookee->meleerange + lookee->radius)
+			if (mindist || distance_squared > ( (lookee->meleerange + lookee->radius) * (lookee->meleerange + lookee->radius) ) )
```
Comment (edward-san) [inline src/playsim/p_enemy.cpp:1188]: this one is wrong here, see all the suggestions which take care of `dist > value`.

File: src/playsim/p_enemy.cpp (hunk 7 above)
Comment (edward-san) [inline src/playsim/p_enemy.cpp:1935]: wrong here

File: src/playsim/p_enemy.cpp (hunk 8 above)
Comment (edward-san) [inline src/playsim/p_enemy.cpp:2006]: partially wrong here

File: src/playsim/p_map.cpp [hunk 15]
```diff
@@ -2065,15 +2065,20 @@ void P_FakeZMovement(AActor *mo)
...
+			//still need delta intact, because it can be negative
+			//and square of a real number always positive
 			double delta = mo->target->Center() - mo->Z();
-			if (delta < 0 && dist < -(delta * 3))
+			//why it multiplies by 3 in original function?
+			double delta_squared = delta * abs(delta) * 3 * 3;
+			if (delta_squared < 0 && dist_squared < -(delta_squared) )
 				mo->AddZ(-mo->FloatSpeed);
-			else if (delta > 0 && dist < (delta * 3))
+
```
Comment (edward-san) [inline src/playsim/p_map.cpp:2076]: please drop this empty line

File: src/playsim/p_actionfunctions.cpp (hunk 10 above)
Comment (MekBoss) [reply src/playsim/p_actionfunctions.cpp:2464]: Ehhhh....but its literally same thing as before, just in another form.

File: src/playsim/p_actionfunctions.cpp (hunk 12 above)
Comment (MekBoss) [reply src/playsim/p_actionfunctions.cpp:2556]: Same here

File: src/playsim/p_enemy.cpp (hunk 14 above)
Comment (MekBoss) [reply src/playsim/p_enemy.cpp:1188]: Okay.

File: src/playsim/p_enemy.cpp (hunk 7 above)
Comment (MekBoss) [reply src/playsim/p_enemy.cpp:1935]: Okay(2)

File: src/playsim/p_map.cpp (hunk 15 above)
Comment (MekBoss) [reply src/playsim/p_map.cpp:2076]: I wanted to group "different" operations in one "block", but Okay(3)

File: src/playsim/p_actionfunctions.cpp (hunk 10 above)
Comment (edward-san) [reply src/playsim/p_actionfunctions.cpp:2464]: Wrong: if `dist_max` is negative, then the condition in your code returns `false`, which is not what the original code did (in that case it returns `true`). You need to preserve the original behavior here. The same is said for the other comments.

**Comment by coelckers** [comment]
TBH, this is way too risky for me. As an entire catch-all for this issue this PR is simply too big, too invasive and too dangerous for something that's a very minor performance gain at the very best.

**Comment by drfrag666** [comment]
Well this is getting a bit confusing but if performance difference is noticeable in real life and now that @edward-san has taken the time to review this i'm definitely interested. So we could merge this in LZDoom in my master branch and see how it works

**Comment by coelckers** [comment]
Even then, you are saddling yourself with one big commit. If this was split into smaller fragments that can be individually reviewed it has a far better chance to make it in, but in its current state I cannot review it because it's too much in too many places.

**Comment by MajorCooke** [comment]
So you want it split up into multiple commits? If so, I could get behind that.

**Comment by coelckers** [comment]
Essentially yes. Some parts would be easy to add, but others may require more caution.

**Comment by MajorCooke** [comment]
Sounds good. I'll do one with the simpler stuff first, then add another (if not several) with the more cautious stuff.

**Comment by drfrag666** [comment]
But that will require more work and @edward-san  has reviewed it and proposed changes, seems @MekBoss  is not going to make the changes so i could merge this as is and then Edoardo could push his modifications if he agrees. I think i'd create a new branch from my master then.

**Comment by MajorCooke** [comment]
Did Mek address all those changes? If so, I can just copy over the work he did and send in the commit chopped up. If not I can add those in myself.

But if Graf agrees to that, then fine by me.

**Comment by edward-san** [comment]
I could do that, but rewritting the history with smaller commits sounds like a better idea, IMHO. If it is acceptable, I can do that instead (with @MekBoss kept as the author).

**Comment by drfrag666** [comment]
Yeah do whatever you think it's better, if you decide to make it in LZDoom just remember that modern LZDoom is my master branch and not the default branch for now.

**Comment by MajorCooke** [comment]
That would be appreciated. I don't know how to keep another as an author.

**Comment by drfrag666** [comment]
That's easy in TortoiseGit in the commit dialog (set author) but i found it only recently, in the Git command line you also can change the commiter but it's trickier.

**Comment by edward-san** [comment]
Pushed some of the changes which I think are okay now. The rest will be taken care of when I have the time.

**Comment by drfrag666** [comment]
Thanks! I've seen it's here: https://github.com/drfrag666/gzdoom/commits/distance_squared

**Comment by MekBoss** [comment]
@drfrag666  its not that I dont want push requested changes, I just was busy with real life things. So, should I redo it, and pull all changes one by one or just drop it or you do it?

**Comment by coelckers** [comment]
If you want to do it, feel free to. The only thing to look out for is to keep the scope of each PR focussed so that they can be reviewed quickly.

**Comment by drfrag666** [comment]
@edward-san is already working on it but he'll be busy with real life for a couple of weeks AFAIR. When it's done i'll release a devbuild for testing.

**Comment by edward-san** [comment]
Hello!

I'm still not fully available due to IRL stuff. Apparently the author is making new PRs with the separated commits as requested, so it seems it is not necessary for me to work on this anymore, is it? 😅

**Comment by drfrag666** [comment]
May be not, but are they the same thing or something completely different? And then would they clash with this stuff? :)
